### PR TITLE
feat(envs): unify world models backend

### PR DESCRIPTION
**Pull request by ZJLi2013**
﻿### Description

PR 1~3 of the plan in #1511: 

1. `WorldModelBackend` is a Protocol — `open_session` / `generate` / `close_session` / `offload` / `onload` — with `WanBackend` and `OpenSoraBackend` as the in-process implementations. The session owns the condition window, so the env hands it over once instead of re-sending five 256×256 frames per env per chunk, and a remote backend gets the trajectory boundaries it would need to cache. Window semantics are unchanged.

2.  `WanEnv` and `OpenSoraEnv` shared an 850-line episode loop almost line for line. What is left of each is `_build_backend` and `_load_reward_model` over a shared `WorldModelEnv`.

3. **Auto reset restarted every slot when any episode ended** (bug reported #1516). With 16 envs, one estimated success pushed all 16 back to a freshly sampled episode while the other 15 still reported `done=False`, in this pr make `elapsed_steps` for per env-slot in `BaseWorldEnv`, `reset()` takes `env_idx`, and `_handle_auto_reset` passes the indices from `dones`.

### Motivation and Context

**The goal is to serve the world model independently of the env**, eventually from another engine. That needs an interface, and this PR is what it takes to have one. The auto-reset bug #1516 comes along because a session is keyed by trajectory, and "which trajectory is on this slot" had no answer while every termination restarted everything.

### How has this been tested?

2×MI300X (ROCm 6.4). Both backends ran end to end with the same shape: 16 envs × 16 steps, two chunks, placement `2-3`, only `env.train` swapped.

- **Unit, 18 passed**, all stubbing their heavy deps so they run on the plain openvla-oft venv without a GPU or a checkpoint. `test_wan_backend.py` (5) and `test_opensora_backend.py` (5) cover session bookkeeping and each window's contents; `test_world_model_autoreset.py` (8) covers per-slot `elapsed_steps`, a subset reset leaving the other slots untouched, the shared time axis keeping its length, and `_handle_auto_reset(dones=[F, T, F, T])` restarting exactly slots 1 and 3.
- **Wan**, `wan_libero_spatial_grpo_openvlaoft` with `RLinf-Wan-LIBERO-Spatial`: exit 0, Global Step 1/1, `env_interact_step` 202.2 s against 201.7–210.2 s over six earlier runs of the same setup.
- **OpenSora**, `opensora_libero_spatial_grpo_openvlaoft` with `RLinf-OpenSora-LIBERO-Spatial`: exit 0, Global Step 1/1, `env_interact_step` 137.8 s against 137.7 / 137.9 s before.
- **`episode_len=16.0` on both runs** shows the per-slot `elapsed_steps` did not move where truncation fires: all slots still reach `max_episode_steps` together.

OpenSora on AMD GPU need 2 patches, will add PR later, The CUDA path is unaffected, so embodied-openvlaoft-opensora-test in CI is still worth running.

### Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)

### Checklist

- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: rlinf/envs/world_model/wan_backend.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        env_ids: Sequence[int],
+        init_frames: FrameQueue,
+        task_ids: Sequence[Any],
+        seeds: Sequence[int],
+    ) -> None:
+        # init_frames are not retained: the pipeline re-conditions from the window the env sends with
+        # every generate. A pooled backend would encode them here.
+        for env_id, task_id, seed in zip(env_ids, task_ids, seeds):
+            self._sessions[int(env_id)] = {"task_id": task_id, "seed": int(seed)}
+
```
[the remark this answers is not in the document]
Comment (ZJLi2013) [reply rlinf/envs/world_model/wan_backend.py]: applied

**Comment by ZJLi2013** [comment]
> ## Pull request overview
> This PR introduces a backend interface layer for the Wan world-model environment so frame generation can be routed through interchangeable “backends” (in-process now, with room for remote/served implementations later), while keeping episode semantics (reset/auto-reset/rewards/metrics) inside `WanEnv`.
> 
> **Changes:**
> 
> * Added a `WorldModelBackend` Protocol defining a session-shaped generation interface (`open_session` / `generate` / `close_session` / `offload` / `onload`).
> * Moved Wan pipeline construction + generation into a new in-process `WanBackend`.
> * Refactored `WanEnv` to use `WanBackend` and added unit tests for backend session/seed behavior.
> 
> ### Reviewed changes
> Copilot reviewed 4 out of 4 changed files in this pull request and generated 2 comments.
> 
> File	Description
> `rlinf/envs/world_model/backend.py`	Introduces the `WorldModelBackend` Protocol and `FrameQueue` type alias to define the env↔backend seam.
> `rlinf/envs/world_model/wan_backend.py`	New in-process backend encapsulating `WanVideoPipeline`, session bookkeeping, and generation/offload/onload.
> `rlinf/envs/world_model/world_model_wan_env.py`	Refactors `WanEnv` to open/close sessions on reset and delegate frame generation/offload/onload to the backend.
> `tests/unit_tests/test_world_model_backend.py`	New unit tests validating session bookkeeping and the “single seed per batch” constraint.
> Suppressed comments (2)
> **rlinf/envs/world_model/wan_backend.py:93**
> 
> * _pipe_kwargs() builds the condition batch from `condition` but uses `env_ids` for session/seed lookup. If these ever differ in length (or get out of sync), you'll generate a batch with mismatched rows vs. seed/session metadata. Add a length check and iterate `condition` in lockstep with `env_ids`.
> 
> ```
>         batch_input_image = []
>         batch_input_image4 = []
>         for frames in condition:
>             imgs = []
>             for frame in frames:
> ```
> 
> **rlinf/envs/world_model/wan_backend.py:127**
> 
> * generate() iterates `range(len(condition))` when converting pipeline output to tensors. If `condition` length ever diverges from the actual batch size (`env_ids`), this will index the wrong rows (or crash). Iterate by `env_ids` batch size instead.
> 
> ```
>         videos = []
>         for env_idx in range(len(condition)):
>             frames = []
> ```
> 
> 💡 [Add a `code-review` agent skill](/RLinf/RLinf/new/main?filename=.github/skills/code-review/SKILL.md) or configure MCP servers for context-aware, tailored reviews. [Learn more in the docs.](https://docs.github.com/en/copilot/how-tos/use-copilot-agents/request-a-code-review/use-code-review#mcp-servers-and-agent-skills)

please review again, expand wan to opensora, and add per env slot for world model as env

File: tests/unit_tests/test_opensora_backend.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
+# Copyright 2026 The RLinf Authors.
```
Comment (Lin-xs) [inline tests/unit_tests/test_opensora_backend.py]: Please delete all unit tests.

File: rlinf/envs/sim/world_model/opensora_backend.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+
+class OpenSoraBackend:
+    """In-process backend holding the action-conditioned OpenSora STDiT and its VAE.
+
+    A session holds the trajectory's condition window as a queue of latents, so pixels are
+    only ever decoded on the way out.
+    """
+
+    def __init__(self, cfg, device: torch.device, device_str: str):
```
Comment (Lin-xs) [inline rlinf/envs/sim/world_model/opensora_backend.py]: Could you please merge `device` and `device_str`?

File: rlinf/envs/sim/world_model/world_model_opensora_env.py [hunk 4]
```diff
@@ -12,869 +12,38 @@
...
-            chunk_terminations,
-            chunk_truncations,
-            [infos],
+        rm_cfg = OmegaConf.to_container(
+            self.cfg.world_model_cfg.reward_model, resolve=True
         )
+        return build_module(rm_cfg, MODELS)
 
 
 # PYTHONPATH="/mnt/project_rlinf/jzn/workspace/opensora:$PYTHONPATH" python -m rlinf.envs.sim.world_model.world_model_opensora_env
```
Comment (Lin-xs) [inline rlinf/envs/sim/world_model/world_model_opensora_env.py]: Please delete this absolute path.

File: rlinf/envs/sim/world_model/world_model_opensora_env.py [hunk 5]
```diff
@@ -12,869 +12,38 @@
...
 if __name__ == "__main__":
```
Comment (Lin-xs) [inline rlinf/envs/sim/world_model/world_model_opensora_env.py]: Will this file be executed separately? If yes, please extract the code under `if __name__ == "__main__"` into a file under `examples/`, and add the usage steps to the docs.

File: rlinf/envs/sim/world_model/world_model_wan_env.py [hunk 6]
```diff
@@ -12,789 +12,51 @@
...
+        if self.cfg.reward_model.type != "TaskEmbedResnetRewModel":
+            return None
+        # One instruction per scored frame, so each description repeats over its chunk
+        instructions = []
+        for env_idx in range(self.num_envs):
+            instructions.extend([self.task_descriptions[env_idx]] * self.chunk)
+        return instructions
 
 
 # PYTHONPATH="/mnt/project_rlinf/jzn/workspace/release/DiffSynth-Studio:$PYTHONPATH" python -m rlinf.envs.sim.world_model.world_model_wan_env
```
Comment (Lin-xs) [inline rlinf/envs/sim/world_model/world_model_wan_env.py]: Same issue here.

File: rlinf/envs/sim/world_model/world_model_env.py [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+        past_dones = torch.logical_or(past_terminations, past_truncations)
+
+        if past_dones.any() and self.auto_reset:
+            extracted_obs, infos = self._handle_auto_reset(
+                past_dones, extracted_obs, {}
+            )
+        else:
+            infos = {}
+
+        infos = self._record_metrics(
```
Comment (Lin-xs) [inline rlinf/envs/sim/world_model/world_model_env.py]: This path auto-resets before recording metrics, which is the reverse of the `LIBERO` env. Is that expected, or should the order match `LIBERO`?

File: tests/unit_tests/test_opensora_backend.py (hunk 2 above)
Comment (ZJLi2013) [reply tests/unit_tests/test_opensora_backend.py]: sure,

File: rlinf/envs/sim/world_model/world_model_opensora_env.py (hunk 5 above)
Comment (ZJLi2013) [reply rlinf/envs/sim/world_model/world_model_opensora_env.py]: delete this

File: rlinf/envs/sim/world_model/world_model_env.py (hunk 7 above)
Comment (ZJLi2013) [reply rlinf/envs/sim/world_model/world_model_env.py]: should match libero, fixed

**Comment by ZJLi2013** [comment]
hi, @Lin-xs  thanks for review, all sugggestion updated

**Comment by andylin-hao** [comment]
Hi @ZJLi2013, could you run a pre-commit code lint pass? The linter check failed.

**Comment by ZJLi2013** [comment]
> Hi @ZJLi2013, could you run a pre-commit code lint pass? The linter check failed.

fixed, thanks

**Review by andylin-hao — APPROVED** [review]

**Comment by guozhen1997** [comment]
Hi @ZJLi2013, could you please review and merge this PR https://github.com/ZJLi2013/RLinf/pull/1 ?
This PR unifies world-model envs behind a single `WorldModelEnv`, selected by env_type: world_model and backend: wan / opensora. Generation lives in registered backends so Wan and OpenSora stay optional dependencies, and adding a model means implementing `WorldModelBackend` and registering it rather than adding another env class or `get_env_cls` branch.

**Comment by ZJLi2013** [comment]
> Hi @ZJLi2013, could you please review and merge this PR [ZJLi2013#1](https://github.com/ZJLi2013/RLinf/pull/1) ? This PR unifies world-model envs behind a single `WorldModelEnv`, selected by env_type: world_model and backend: wan / opensora. Generation lives in registered backends so Wan and OpenSora stay optional dependencies, and adding a model means implementing `WorldModelBackend` and registering it rather than adding another env class or `get_env_cls` branch.

very good design. thanks a lot, merged

**Review by guozhen1997 — APPROVED** [review]
